presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,283 @@
+                    .collect(toImmutableList());
+            return new JoinNode(
+                    idAllocator.getNextId(),
+                    INNER,
+                    left,
+                    right,
+                    flipJoinCriteriaIfNecessary(joinConditions, left),
+                    outputSymbols,
+                    joinFilters.isEmpty() ? Optional.empty() : Optional.of(ExpressionUtils.and(joinFilters)),
+                    Optional.empty(),
+                    Optional.empty(),
+                    Optional.empty());
+        }
+
+        private static List<JoinNode.EquiJoinClause> flipJoinCriteriaIfNecessary(List<JoinNode.EquiJoinClause> joinCriteria, PlanNode left)
+        {
+            return joinCriteria
+                    .stream()
+                    .map(criterion -> left.getOutputSymbols().contains(criterion.getLeft()) ? criterion : criterion.flip())
+                    .collect(toImmutableList());
+        }
+
+        private static PlanNode getJoinSource(PlanNodeIdAllocator idAllocator, List<PlanNode> nodes, Expression filter, List<Symbol> outputSymbols)
+        {
+            PlanNode planNode;
+            if (nodes.size() == 1) {
+                planNode = getOnlyElement(nodes);
+                if (!(BooleanLiteral.TRUE_LITERAL).equals(filter)) {
+                    return new FilterNode(idAllocator.getNextId(), planNode, filter);
+                }
+                return planNode;
+            }
+            return new JoinGraphNode(idAllocator.getNextId(), nodes, filter, outputSymbols);
+        }
+
+        private JoinNode chooseJoinOrder(JoinGraphNode joinGraph)
+        {
+            JoinNode join = memo.get(joinGraph);
+            if (join == null) {
+                join = generatePartitions(joinGraph.getSources().size())
+                        .map(partitioning -> createJoinNodeTree(joinGraph, partitioning, idAllocator))
+                        .min(planNodeOrdering)
+                        .orElseThrow(() -> new IllegalStateException("joinOrders cannot be empty"));
+                memo.put(joinGraph, join);
+            }
+            return join;
+        }
+
+        private JoinNode createJoinNodeTree(JoinGraphNode joinGraph, Set<Integer> partitioning, PlanNodeIdAllocator idAllocator)
+        {
+            JoinNode placeholderJoin = createJoinAccordingToPartitioning(joinGraph, partitioning, idAllocator);
+            PlanNode left = placeholderJoin.getLeft();
+            PlanNode right = placeholderJoin.getRight();
+            if (left instanceof JoinGraphNode) {
+                left = chooseJoinOrder((JoinGraphNode) left);
+            }
+            if (right instanceof JoinGraphNode) {
+                right = chooseJoinOrder((JoinGraphNode) right);
+            }
+
+            JoinNode joinNode = (JoinNode) placeholderJoin.replaceChildren(ImmutableList.of(left, right));
+            return chooseJoinNodeProperties(joinNode);
```

Comment:
After second though, I realized I don't know how would I go implemented that. It wouldn't be just "search a bit more possible plans", rather it would be different recursion code altogether.

Right, don't do that at this point.